overlord/snapstate/snapstate.go

```diff
@@ -1361,17 +1365,95 @@ var ValidateRefreshes func(st *state.State, refreshes []*snap.Info, ignoreValida
 // UpdateMany updates everything from the given list of names that the
 // store says is updateable. If the list is empty, update everything.
 // Note that the state must be locked by the caller.
-func UpdateMany(ctx context.Context, st *state.State, names []string, userID int, flags *Flags) ([]string, []*state.TaskSet, error) {
-	return updateManyFiltered(ctx, st, names, userID, nil, flags, "")
+func UpdateMany(ctx context.Context, st *state.State, names []string, revOpts []*RevisionOptions, userID int, flags *Flags) ([]string, []*state.TaskSet, error) {
+	return updateManyFiltered(ctx, st, names, revOpts, userID, nil, flags, "")
 }
 
 // EnforceSnaps installs/updates/removes snaps reported by validationErrorToSolve.
 // validationSets is the list of sets passed by the user and it's used in the
 // final stage to update validation-sets tracking in the state.
 // userID is used for store auth.
-func EnforceSnaps(ctx context.Context, st *state.State, validationSets []string, validationErrorToSolve *snapasserts.ValidationSetsValidationError, userID int) (tasksets []*state.TaskSet, names []string, err error) {
-	// TODO
-	return nil, nil, fmt.Errorf("not implemented")
+func EnforceSnaps(ctx context.Context, st *state.State, validationSets []string, valErr *snapasserts.ValidationSetsValidationError, userID int) ([]*state.TaskSet, []string, error) {
+	if len(valErr.InvalidSnaps) != 0 {
+		invSnaps := make([]string, 0, len(valErr.InvalidSnaps))
+		for invSnap := range valErr.InvalidSnaps {
+			invSnaps = append(invSnaps, invSnap)
+		}
+		return nil, nil, fmt.Errorf("cannot refresh enforce validation sets that requires removing snaps %s", strutil.Quoted(invSnaps))
+	}
+
+	// use the same lane for installing and refreshing so everything is reversed
+	lane := st.NewLane()
+
+	var affected []string
+	var tasksets []*state.TaskSet
+	var prev *state.TaskSet
+	addTaskset := func(tss []*state.TaskSet) {
+		for _, ts := range tss {
+			if prev != nil {
+				ts.WaitAll(prev)
```

Comment:
This is not how newChange in deamon usually consumes the results of InstallMany/UpdateMany.
I don't think we can just do this, both InstallMany and UpdateMany might have their opinion on how some snap updates/install depends on other, so I fear we might in general adding loops or something unexpected here. We really just need a wait between the different phases, not add extra waits between the TaskSets returned by the same functions. My feeling is also that we should updates before we do installs.